src/features/world/scenes/BaseScene.ts

```diff
@@ -9,24 +9,98 @@ import { INITIAL_BUMPKIN } from "features/game/lib/constants";
 import { BumpkinContainer } from "../containers/BumpkinContainer";
 import { interactableModalManager } from "../ui/InteractableModals";
 import {
+  ChatContext,
   ChatMessageReceived,
   MachineInterpreter,
   PlayerJoined,
   PlayerQuit,
+  RoomEvent,
+  RoomId,
 } from "../roomMachine";
 import { CONFIG } from "lib/config";
 import { NPCName, NPC_WEARABLES } from "lib/npcs";
 import { npcModalManager } from "../ui/NPCModals";
 import { BumpkinParts } from "lib/utils/tokenUriBuilder";
 import { SUNNYSIDE } from "assets/sunnyside";
+import { EventObject, State } from "xstate";
 
 export type NPCBumpkin = {
   x: number;
   y: number;
   npc: NPCName;
 };
 
-export class BaseScene extends Phaser.Scene {
+export abstract class BaseScene extends Phaser.Scene {
+  abstract roomId: RoomId;
+  eventListener: (event: EventObject) => void;
+  transitionListener: (
+    state: State<ChatContext, RoomEvent, any, any, any>,
+    event: RoomEvent
+  ) => void;
+
+  constructor(key: string) {
+    super(key);
+
+    this.eventListener = (event) => {
+      if (event.type === "CHAT_MESSAGE_RECEIVED") {
+        const { sessionId, text } = event as ChatMessageReceived;
+        if (
+          sessionId &&
+          String(sessionId).length > 4 &&
+          this.playerEntities[sessionId]
+        ) {
+          this.playerEntities[sessionId].speak(text);
+        }
+      }
+
+      if (event.type === "PLAYER_JOINED") {
+        const { sessionId, x, y, clothing, roomId } = event as PlayerJoined;
+
+        if (roomId !== this.roomId) return;
```

Comment:
Events now receive a room ID. This allows scenes to filter events that are not for them.

The alternative to this is to deregister the listeners, but I think allowing filtering is a bit more flexible.

Reply:
Since we deregistered the listeners above in what scenario does this get called?

Reply:
The listener doesn't get de-registered when the scene is shutdown. That deregistration call only happens during `create` - which would be called when the scene is restarted. It's only really there to ensure `onEvent` doesn't get attached twice.